lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -455,6 +463,7 @@ defmodule Mix.Compilers.Elixir do
     source =
       source(
         source,
+        digest: digest(file),
```

Comment:
Your change in `update_stale_sources` means if a file is recompiled but not changed, you are keeping the digest, so I think we can leverage it here. I will push this change and see how the test suite behaves. :)

```suggestion
        # We preserve the digest if the file is recompiled but not changed
        digest: source(source, :digest) || digest(file),
```